src/core/bst.h

```diff
@@ -188,361 +133,540 @@ class BST {
     BST(BST_TYPE::BST_TYPE type) : type(type), lineno(0) {}
 #endif
     BST(BST_TYPE::BST_TYPE type, uint32_t lineno) : type(type), lineno(lineno) {}
-};
-
-class BST_expr : public BST {
-public:
-    virtual void* accept_expr(ExprVisitor* v) = 0;
 
-    BST_expr(BST_TYPE::BST_TYPE type) : BST(type) {}
-    BST_expr(BST_TYPE::BST_TYPE type, uint32_t lineno) : BST(type, lineno) {}
+    virtual bool isBST_dst() const { return false; }
 };
 
 class BST_stmt : public BST {
 public:
     virtual void accept_stmt(StmtVisitor* v) = 0;
 
-    int cxx_exception_count = 0;
-
     BST_stmt(BST_TYPE::BST_TYPE type) : BST(type) {}
+    BST_stmt(BST_TYPE::BST_TYPE type, uint32_t lineno) : BST(type, lineno) {}
 };
 
-class BST_slice : public BST {
+// base class of all nodes which have a single destination vreg
+class BST_dst : public BST_stmt {
 public:
-    virtual void* accept_slice(SliceVisitor* s) = 0;
-    BST_slice(BST_TYPE::BST_TYPE type) : BST(type) {}
-    BST_slice(BST_TYPE::BST_TYPE type, uint32_t lineno) : BST(type, lineno) {}
+    int vreg_dst = VREG_UNDEFINED;
+    BST_dst(BST_TYPE::BST_TYPE type) : BST_stmt(type) {}
+    BST_dst(BST_TYPE::BST_TYPE type, int lineno) : BST_stmt(type, lineno) {}
+
+    bool isBST_dst() const override { return true; }
 };
 
-class BST_Name;
 
-class BST_arguments : public BST {
+
+class BST_Assert : public BST_stmt {
 public:
-    // no lineno attributes
-    std::vector<BST_expr*> defaults;
+    int vreg_msg = VREG_UNDEFINED;
 
     virtual void accept(BSTVisitor* v);
+    virtual void accept_stmt(StmtVisitor* v);
 
-    BST_arguments() : BST(BST_TYPE::arguments) {}
+    BST_Assert() : BST_stmt(BST_TYPE::Assert) {}
 
-    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::arguments;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::Assert;
 };
 
-class BST_Assert : public BST_stmt {
+class BST_UnpackIntoArray : public BST_stmt {
 public:
-    BST_expr* msg, *test;
+    int vreg_src = VREG_UNDEFINED;
+    const int num_elts;
+    int vreg_dst[1];
 
     virtual void accept(BSTVisitor* v);
     virtual void accept_stmt(StmtVisitor* v);
 
-    BST_Assert() : BST_stmt(BST_TYPE::Assert) {}
+    static BST_UnpackIntoArray* create(int num_elts) {
+        BST_UnpackIntoArray* o
+            = (BST_UnpackIntoArray*)new char[offsetof(BST_UnpackIntoArray, vreg_dst) + num_elts * sizeof(int)];
+        new (o) BST_UnpackIntoArray(num_elts);
+        return o;
+    }
 
-    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::Assert;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::UnpackIntoArray;
+
+private:
+    BST_UnpackIntoArray(int num_elts) : BST_stmt(BST_TYPE::UnpackIntoArray), num_elts(num_elts) {
+        for (int i = 0; i < num_elts; ++i) {
+            vreg_dst[i] = VREG_UNDEFINED;
+        }
+    }
 };
 
-class BST_Assign : public BST_stmt {
+class BST_CopyVReg : public BST_dst {
 public:
-    BST_expr* target;
-    BST_expr* value;
+    int vreg_src = VREG_UNDEFINED; // this vreg will not get killed!
```

Comment:
I think it'd be great to have a comment saying why we need a CopyVReg bytecode.  And maybe have a comment towards the top of the file that goes over the properties of the bytecode, such as that source vregs are always killed after use, how constants are represented, etc.